### PR TITLE
Add a system for modifying entity names without causing conflicts

### LINKED ISSUE #27670: Entity rename mechanics are incompatible with each other
## Description

There is no proper support for renaming entities in a way where multiple systems are compatible with each other. Examples of these systems include gluing/lubing, labeling, metamorphic glass names and pAIs. Conflicts between these systems can cause an item to lose one name change if another system changes it under certain circumstances.

As an example, the label system saves the pre-label name in the entity's LabelComponent. This is the name that the entity gets reset to when the label is removed. That means any other system that changes an object name gets overwritten if the label is removed:

- You can remove the pAI's name and return it to "personal ai device" by labeling it before a role joins, and then removing the label after.
- You can remove glue and lube tags from an object by labeling it before applying the substance and then unlabeling it after.
- Depending on implementation, this can happen in reverse.
  - For glue/lube, if you apply the substance to a labeled object and then remove the label, the object gets the label back when the effect wears off.
  - For pAIs the label gets hidden when a pAI role joins and remains hidden once the role leaves as the entity gets reset to its prototype name.

Metamorphic glasses is another system that has this same problem.

**Reproduction**

1. Apply a label (e.g. "test") to an object.
  - Resulting name: "object (test)"
2. Apply space glue to the object. 
  - Resulting name: "Glued object (test)"
3. Remove the label (by leaving the label text field blank). 
  - Expected name: "Glued object"
  - Resulting name: "object"

**Screenshots**

image

Example of how a metamorphic glass can be disguised using this method by applying a label while being a beer glass, changing its contents and then removing the label.

**Additional context**

Found in PR #26749

### PR DESCRIPTION
**Pull request by Tayrtahn**
Requires space-wizards/RobustToolbox#5216

## About the PR

Adds NameModifierSystem, which aims to fix the problems that arise from different systems renaming entities.

Modifies a bunch of systems that previously could cause renaming conflicts when used on the same entity. Specifically, this fixes issues with metamorphic glasses, glue, lube, flares/glowsticks, forensic pads, DNA scrambler implants, PAIs, baby animals, borgs, zombies, cluwnes, chameleon clothing, and the admin renaming command and verbs.

Also includes ModifyWearerName Component/System both as a demo and to make it easier to add clothing that does simple name modifications. It's not used on anything in this PR though.

## Why / Balance

See #27670 for an explanation of the issues that can come up with multiple systems renaming things. In short, the old method for renaming an entity was to copy the old name into a component field and then copy it back to revert the name. This meant that any additional changes to the name between those actions would be bulldozed. This system prevents that and makes it easy to add and remove name modifications. You can now have a "glued lubed mojito glass (for nerds)" and refill it to make a "glued lubed screwdriver glass (for nerds)" without issues.

Fixes #27670

## Technical details

When `NameModifierSystem.RefreshNameModifiers` is called, a `RefreshNameModifiersEvent` is passed around the entity and any subscribed components can add modifiers to to it. The event is also relayed to equipped items.

A bunch of existing systems that temporarily changed entity names have been modified to use NameModifierSystem instead. Systems that make permanent changes still work and will have their changes respected by this system.

This PR also includes ModifyWearerNameComponent/System, but doesn't use it on anything. I used it quite a bit in testing, and it nicely demonstrates the integration with InventoryRelayEvent. It's a simple component that can be added to clothing prototypes to make them add a prefix, postfix, or override to the wearer's name while equipped. For example, the lawyer's badge can add the "Attorney" title to a wearer's name with just a little yml:
```yml
- type: ModifyWearerName
  locId: lawyer-name-prefix
```

## Media

Multiple modifiers surviving changes:
Screenshot 2024-05-06 at 5 47 22 PM
Screenshot 2024-05-06 at 5 49 15 PM

This may not make a lot of sense, but at least the name behaves properly:
Screenshot 2024-05-06 at 5 50 45 PM

ModifyWearerNameComponent:

https://github.com/space-wizards/space-station-14/assets/85356/e395a24a-5025-440a-b131-36357f881ff6

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

Systems that apply reversible changes to entity names should be changed to use NameModifierSystem so they don't conflict with other name changes.

**Changelog**

Nah.

### CODE REVIEW

**Comment by Tayrtahn** [comment]
Just realized that specifying prefix/postfix in the code is bad for localization, since what might be a prefix in one language can be a postfix in another ("lubricated crowbar" -> "palanca lubricada").

I'll rework this to ditch the prefix/postfix concept and use Loc strings for formatting.

**Comment by Tayrtahn** [comment]
Did a pretty major overhaul of this. Gave it a new name, ditched the prefix/postfix/override concept in favor of modifiers, and made it play a lot more nicely with other systems - permanent renaming an entity no longer has to manually call this system and can just use the normal `MetaDataSystem.SetEntityName` method.

This new way of working does require an engine PR to add an event for renaming entities which this subscribes to.

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (AJCM-git) [inline Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:110]: it may be cleaner to use the params keyword here

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (AJCM-git) [inline Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:127]: this is kinda odd, if this is supposed to always override the lesser priority modifiers with the higher priority ones, couldnt you just order by priority (extra points without linq) and just use the last element of the sorted list instead of unnecessarily looping?

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (AJCM-git) [inline Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:45]: is this resolve here doing anything when you just add the component if it doesnt have it anyways? (the xmldoc of this method could be slightly clear about this fact too, it implies it but doesnt say it directly)

**Review by AJCM-git — CHANGES_REQUESTED** [review]
This is looking good, just some details, also I think master RT submodule still doesnt have the required merged pr
Also, remember to write something for codebase changes after this is merged

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (Tayrtahn) [reply Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:127]: This is a little tricky - it needs to loop so that modifiers that work as a prefix/postfix can be applied together.

So if you have a metamorphic glass with beer in it and you put glue, lube, and a label on it, the different systems will add `glued {$baseName}` (priority 0),  `lubed {$baseName}` (priority 0), `beer glass` (priority -1), and `{$baseName} (label)`

This code loops through like: 
base name: "metamorphic glass"
`beer glass` -> "beer glass"
`glued {$baseName}` -> "glued beer glass"
`lubed {$baseName}` -> "lubed glued beer glass"
`{$baseName} (label)` -> "lubed glued beer glass (label)"

So the loop is necessary. I definitely can add some comments to make the logic clearer for future maintainers.

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (Tayrtahn) [reply Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:45]: Surprisingly, it does seem to be necessary. Without it, another system calling this method must check for the component and pass it in if it exists, which is kinda terrible.

If they don't, this code will see ent.Comp as null, even though the entity does have the component. That causes all kinds of problems, with BaseName getting set to include modifiers, causing them to get permanently added to the name, and to stack up with repeated refreshes - in testing, I had a glued glued glued glued crowbar that wasn't actually glued.

The alternative would be to use EnsureComp at the top of the method and then RemComp later if the component isn't needed, but I assume that adding and removing a new component is less efficient than calling Resolve.

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (AJCM-git) [reply Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:127]: Ah no i kinda got that, i think what i misunderstood is what _modifiers was actually getting populated with, kinda conflated it with the extra args, my bad

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (AJCM-git) [inline Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:141]: I see what you meant with more allocations, but cant you make the _modifiers.ExtraArgs be an array and do something like 
```c#
        foreach (var (locId, _, extraArgs) in _modifiers.OrderBy(n => n.Priority))
        {
            // Grab any extra args needed by the Loc string
            var args = modifier.ExtraArgs ?? [];
            // Add the current version of the entity name as an arg
            Array.Resize(ref args, args.Length + 1);
            // Resolve the Loc string and use the result as the base in the next iteration.
            name = Loc.GetString(modifier.LocId, args);
```
you are still making a new array to copy but at least we get rid of 2 linq calls, just something i thought about, could just not work

File: Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs
Comment (AJCM-git) [reply Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs:50]: I mean i guess whats weird is that you arent using the resolve return value so you are basically just doing a TryComp without just directly calling it, so i would either do that or make use of the resolve to the fullest and do something like :
```c#
        var meta = MetaData(entity);
        var baseName = meta.EntityName;
        if (Resolve(entity, ref entity.Comp, logMissing: false))
            baseName = entity.Comp.BaseName;
```

**Review by AJCM-git — APPROVED** [review]
looks aight

**Pull request merged**
